### PR TITLE
Handling Endianness correctly

### PR DESCRIPTION
**Pull request by brdandu**
Handling the negative default values and float values for attributes correctly in terms of endianness(little/big)

### CODE REVIEW

**Comment by tecimovic** [comment]
Some unit test seems to be failing.

**Comment by brdandu** [comment]
> Some unit test seems to be failing.
https://github.com/project-chip/zap/issues/593 Logged this for a UI warning when user enters an invalid hex value as an attribute default which was the reason for the test failure. Will be a better experience for a user to see this in the UI immediately than during compile time.

File: src-electron/generator/helper-zcl.js [hunk 1]
```diff
@@ -2209,6 +2209,78 @@ async function if_mfg_specific_cluster(clusterId, options) {
...
+}
+
+/**
+ *
+ * @param {*} value
+ * @param {*} size
+ * @returns the negative hex value for a given positive hex value padded with
+ * the given size.
+ */
+async function get_negative_hex_value(value, size) {
```
Comment (tecimovic) [inline src-electron/generator/helper-zcl.js]: I thought we generally don't use `get_` for our helpers. We mostly do `as_` or something?

File: src-electron/generator/helper-zcl.js [hunk 2]
```diff
@@ -2209,6 +2209,78 @@ async function if_mfg_specific_cluster(clusterId, options) {
...
+      } else {
+        if (carry == '1') {
+          result = '1' + result
+          carry = '0'
+        } else {
+          result = tempBinary.charAt(i) + result
+        }
+      }
+    }
+    tempBinary = result
```
Comment (tecimovic) [inline src-electron/generator/helper-zcl.js]: Wow. Is there seriously no package out there that does this? This seems very complicated...

File: src-electron/generator/helper-zcl.js [hunk 3]
```diff
@@ -2299,13 +2227,13 @@ async function as_generated_default_macro(value, attributeSize, options) {
     let tempValue = value > 0 ? value : Math.abs(value)
     // Float value
     if (!isNaN(value) && value.toString().indexOf('.') != -1) {
-      temp = await convert_float_to_hex(tempValue, attributeSize)
+      temp = types.convertFloatToHex(tempValue, attributeSize)
```
Comment (tecimovic) [inline src-electron/generator/helper-zcl.js:2230]: Have you purposefully lost `await` here?

Comment (tecimovic) [reply src-electron/generator/helper-zcl.js:2230]: Yeah... Having no await is correct, looking at the actual functions.

**Review by tecimovic — APPROVED** [review]
Minor comments, but can merge.

File: src-electron/generator/helper-zcl.js (hunk 1 above)
Comment (brdandu) [reply src-electron/generator/helper-zcl.js]: fixed it

File: src-electron/util/types.js [hunk 4]
```diff
@@ -78,6 +78,78 @@ async function typeSizeAttribute(db, zclPackageId, at, defaultValue = null) {
...
+ * @param {*} value
+ * @param {*} size
+ * @returns The negative hex value for a given positive hex value padded with
+ * the given size.
+ */
+function convertFloatToHex(value, size) {
+  let res = ''
+  const arrayBuffer = new ArrayBuffer(size)
+  const dataView = new DataView(arrayBuffer)
+  dataView.setFloat32(0, value, false)
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:92]: What if this is a 64-bit float?

Seems like this function should check for `size` being 4 or 8, throw if not one of those, then use the right accessors on the dataview.

File: src-electron/util/types.js [hunk 5]
```diff
@@ -78,6 +78,78 @@ async function typeSizeAttribute(db, zclPackageId, at, defaultValue = null) {
   }
 }
 
+/**
+ *
+ * @param {*} value
+ * @param {*} size
+ * @returns The negative hex value for a given positive hex value padded with
+ * the given size.
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:86]: No, that's not what this function does.

File: src-electron/util/types.js [hunk 6]
```diff
@@ -78,6 +78,78 @@ async function typeSizeAttribute(db, zclPackageId, at, defaultValue = null) {
...
+ */
+function convertFloatToHex(value, size) {
+  let res = ''
+  const arrayBuffer = new ArrayBuffer(size)
+  const dataView = new DataView(arrayBuffer)
+  dataView.setFloat32(0, value, false)
+  for (let i = 0; i < size; i++) {
+    let t = dataView.getUint8(i).toString(16)
+    res += t == 0 ? '00' : t
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:95]: Why is this `t == 0` check here?

In any case, I don't think this is right.  If we get a byte with value `1` this will append `"1"` to the string, whereas it should append `"01"`, no?

File: src-electron/util/types.js [hunk 7]
```diff
@@ -78,6 +78,78 @@ async function typeSizeAttribute(db, zclPackageId, at, defaultValue = null) {
...
+}
+
+/**
+ *
+ * @param {*} value
+ * @param {*} size
+ * @returns the negative hex value for a given positive hex value padded with
+ * the given size.
+ */
+function asNegativeHexValue(value, size) {
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:107]: This function is being used for floats as well, per above, and for floats this makes absolutely no sense....

I am also not sure why we are doing this transformation on the hex-encoded string for integers.  It seems like for integers (at least 1/2/4-byte ones; 8-byte is complicated) we could just detect that the value is negative and convert `2^{1/2/4} + N`, to hex as a positive value.  For 8-byte integers things are more annoying....  But in practice if the actual value is in the [-2^48, 0] range we can do `2^48 + value`, convert to a 6-byte integer, and pad out with 0xFF at the front.  And if it's outside that range, we should probably throw for now.  Not least because if it picks up another few bits we will stop being able to represent it exactly anyway and all this code will go awry, right?

File: src-electron/util/types.js [hunk 8]
```diff
@@ -100,17 +172,34 @@ function longTypeDefaultValue(size, type, value) {
...
-    // First strip off the 0x.
-    if (value.startsWith('0x') || value.startsWith('0X'))
-      value = value.substring(2)
-
-    // Now pad the zeroes to the required size
-    if (size > 0) {
-      while (value.length / 2 < size) {
-        value = '0' + value
+    let temp = ''
+    // Positive or Negative Value
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:176]: There's got to be a way to share more of this code with the code above....

File: test/gen-zigbee.test.js [hunk 9]
```diff
@@ -389,17 +389,17 @@ test(
...
     // Test GENERATED_DEFAULTS for same attribute name but different side of cluster, compare it to above test
     expect(cfgVer2).toContain(
-      '0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0F, 0x0G, 0x0H, /* 28,DEFAULT value for cluster: Green Power, attribute: gp link key, side: client */'
+      '0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0F, 0x0F, 0x0F, /* 28,DEFAULT value for cluster: Green Power, attribute: gp link key, side: client */'
     )
 
     // Test GENERATED_DEFAULTS little endian for attribute of size > 8 is same as big endian. Bytes are not inverted
     expect(cfgVer2).not.toContain(
-      `0x0H, 0x0G, 0x0F, 0x0D, 0x0C, 0x0B, 0x0A, 0x09, 0x08, 0x07, 0x06, 0x05, 0x04, 0x03, 0x02, 0x01, /* 12,DEFAULT value for cluster: Green Power, attribute: gp link key, side: client*/`
+      `0x0F, 0x0F, 0x0F, 0x0D, 0x0C, 0x0B, 0x0A, 0x09, 0x08, 0x07, 0x06, 0x05, 0x04, 0x03, 0x02, 0x01, /* 12,DEFAULT value for cluster: Green Power, attribute: gp link key, side: client*/`
```
Comment (bzbarsky-apple) [inline test/gen-zigbee.test.js:402]: Why is this "12.DEFAULT" while the above toContain has "28,DEFAULT"?  Won't this cause the 12 vs 28 thing to decide what happens no matter what the bytes are?

File: test/helpers.test.js [hunk 10]
```diff
@@ -244,6 +244,32 @@ test(
...
+  },
+  testUtil.timeout.short()
+)
+
+test(
+  'Generated Macro Float Value',
+  () => {
+    let options = { hash: { endian: 'little' } }
+    return zclHelper
+      .as_generated_default_macro('17.0', 8, options)
```
Comment (bzbarsky-apple) [inline test/helpers.test.js:265]: Needs tests for negative doubles.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src-electron/util/types.js (hunk 7 above)
Comment (brdandu) [reply src-electron/util/types.js]: Made changes for this.

File: test/gen-zigbee.test.js (hunk 9 above)
Comment (brdandu) [reply test/gen-zigbee.test.js:402]: The tests are just searching for the content that has been generated. The order in the tests is not the order in which code is generated.

File: src-electron/generator/helper-zcl.js [hunk 11]
```diff
@@ -2233,27 +2233,46 @@ async function if_mfg_specific_cluster(clusterId, options) {
...
   let default_macro_signature = ''
+  let temp = ''
   if (attributeSize > 2) {
-    let default_macro = helperC
-      .asHex(value, null, null)
-      .replace('0x', '')
-      .match(/.{1,2}/g)
+    // String value
+    if (isNaN(value)) {
+      return format_zcl_string_as_characters_for_generated_defaults(
```
Comment (bzbarsky-apple) [inline src-electron/generator/helper-zcl.js:2240]: This is not used for Matter, I assume?  If it is, it's not handling long strings correctly.....

File: src-electron/util/types.js [hunk 12]
```diff
@@ -78,6 +78,77 @@ async function typeSizeAttribute(db, zclPackageId, at, defaultValue = null) {
   }
 }
 
+/**
+ *
+ * @param {*} value
+ * @param {*} size
+ * @returns The big endian value for a given float value padded with
+ * the given size.
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:86]: This should probably document the "0x" prefix, that the value will be hex, etc, etc.

File: src-electron/util/types.js [hunk 13]
```diff
@@ -78,6 +78,77 @@ async function typeSizeAttribute(db, zclPackageId, at, defaultValue = null) {
...
+  }
+  return '0x' + res
+}
+
+/**
+ *
+ * @param {*} value
+ * @param {*} size
+ * @returns The big endian value for a given integer value padded with
+ * the given size.
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:116]: Again, should document the hex and 0x bits.

File: src-electron/util/types.js [hunk 14]
```diff
@@ -78,6 +78,77 @@ async function typeSizeAttribute(db, zclPackageId, at, defaultValue = null) {
...
+ */
+function convertIntToBigEndian(value, size) {
+  const arrayBuffer = new ArrayBuffer(size)
+  const dataView = new DataView(arrayBuffer)
+  if (size == 1) {
+    dataView.setInt8(0, value, false)
+  } else if (size == 2) {
+    dataView.setInt16(0, value, false)
+  } else if (size == 3) {
+    dataView.setInt32(0, value, false)
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:126]: This is not going to do the right thing.  If you're setting as big-endian, this will leave the first byte 0 and set the 2nd through 4th bytes to the value, right?  But then you're going to look at the 1st to 3rd bytes below.

Same thing for sizes 5, 6, 7.

Either we need to do this little-endian, or we need to adjust where we start looking for the value or something.

File: src-electron/util/types.js [hunk 15]
```diff
@@ -78,6 +78,77 @@ async function typeSizeAttribute(db, zclPackageId, at, defaultValue = null) {
...
+  } else if (size == 4) {
+    dataView.setInt32(0, value, false)
+  } else if (size == 5) {
+    dataView.setBigInt64(0, value, false)
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:130]: Are you sure this will work right?  Why would `value` be a BigInt object here?  It almost certainly won't be.....  Is the expectation that `value` is a string and hence will auto-convert to BigInt?

File: src-electron/util/types.js [hunk 16]
```diff
@@ -100,17 +172,34 @@ function longTypeDefaultValue(size, type, value) {
...
-    // Now pad the zeroes to the required size
-    if (size > 0) {
-      while (value.length / 2 < size) {
-        value = '0' + value
+    let temp = ''
+    // Float value
+    if (!isNaN(value) && value.toString().indexOf('.') != -1) {
+      temp = convertFloatToBigEndian(value, size)
+    } else {
+      // Postive value
```
Comment (bzbarsky-apple) [inline src-electron/util/types.js:180]:
```suggestion
      // Positive value
```

File: test/helpers.test.js [hunk 17]
```diff
@@ -244,6 +244,45 @@ test(
   testUtil.timeout.short()
 )
 
+test(
+  'Generated Macro Negative Value',
+  () => {
+    let options = { hash: { endian: 'big' } }
```
Comment (bzbarsky-apple) [inline test/helpers.test.js:250]: Should test all of these in both endian-nesses.

File: src-electron/generator/helper-zcl.js (hunk 11 above)
Comment (brdandu) [reply src-electron/generator/helper-zcl.js:2227]: Yes this is not used by matter. Why do you say that?

Comment (bzbarsky-apple) [reply src-electron/generator/helper-zcl.js:2227]: `format_zcl_string_as_characters_for_generated_defaults` assumes the string length is a singly byte.  For long strings it's two bytes.

File: src-electron/util/types.js (hunk 15 above)
Comment (brdandu) [reply src-electron/util/types.js:130]: I believe this should work. I was mostly thinking in terms of the range. Switching it to setInt32 for a negative integer with attribute size 5 won't work unless I use dataView.setBigInt64

Comment (bzbarsky-apple) [reply src-electron/util/types.js:132]: Sure.  The big question is whether this will throw an exception (which it will if `Value` is a Number: implicit conversion from Number to BigInt is not allowed in JS).  It would be safer to do `BigInt(value)`.

Comment (brdandu) [reply src-electron/util/types.js]: Switched to BigInt

**Review by bzbarsky-apple — APPROVED** [review]
